Cox-Automotive/terraform-provider-alks
# Implement Resource Importer

### Overview
This PR implements the `terraform import resource.resource_name resource_id` functionality. This did not exist for `alks_iamrole` or `alks_iamtrustrole`. Thankfully, we implemented this for the LTK resources, so it follows the exact same strategy. As long as the resources have an ID defined, then the resource can be looked up using the ID and it'll _magically_ work. Tested this in our `examples/` directory! 

### Steps to test:
1. Create an ALKS role in the UI
2. Try to create the same role in TFP
3. Apply, see the error that it already exists.
4. Import the resource by using it's name as the ID.
5. Once imported, re-apply and all should work!

### Example

```
➜  examples git:(implement_import) ✗ tf apply

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # alks_iamrole.test_role will be created
  + resource "alks_iamrole" "test_role" {
      + arn                      = (known after apply)
      + enable_alks_access       = false
      + id                       = (known after apply)
      + include_default_policies = false
      + ip_arn                   = (known after apply)
      + name                     = "TEST-DELETE"
      + role_added_to_ip         = (known after apply)
      + type                     = "AWS CodeBuild"
    }

Plan: 1 to add, 0 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

alks_iamrole.test_role: Creating...

Error: Error creating role: [vx7zpien] Role already exists with the same name: TEST-DELETE

  on alks.tf line 36, in resource "alks_iamrole" "test_role":
  36: resource "alks_iamrole" "test_role" {


➜  examples git:(implement_import) ✗ tf import alks_iamrole.test_role TEST-DELETE
alks_iamrole.test_role: Importing from ID "TEST-DELETE"...
alks_iamrole.test_role: Import prepared!
  Prepared alks_iamrole for import
alks_iamrole.test_role: Refreshing state... [id=TEST-DELETE]

Import successful!

The resources that were imported are shown above. These resources are now in
your Terraform state and will henceforth be managed by Terraform.

➜  examples git:(implement_import) ✗ tf plan
Refreshing Terraform state in-memory prior to plan...
The refreshed state will be used to calculate this plan, but will not be
persisted to local or remote state storage.

alks_iamrole.test_role: Refreshing state... [id=TEST-DELETE]

------------------------------------------------------------------------

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
-/+ destroy and then create replacement

Terraform will perform the following actions:

  # alks_iamrole.test_role must be replaced
-/+ resource "alks_iamrole" "test_role" {
      ~ arn                      = "arn:aws:iam::336692926357:role/acct-managed/TEST-DELETE" -> (known after apply)
        enable_alks_access       = false
      ~ id                       = "TEST-DELETE" -> (known after apply)
      + include_default_policies = false # forces replacement
      + ip_arn                   = (known after apply)
      + name                     = "TEST-DELETE" # forces replacement
      + role_added_to_ip         = (known after apply)
      + type                     = "AWS CodeBuild" # forces replacement
    }

Plan: 1 to add, 0 to change, 1 to destroy.

------------------------------------------------------------------------

Note: You didn't specify an "-out" parameter to save this plan, so Terraform
can't guarantee that exactly these actions will be performed if
"terraform apply" is subsequently run.

➜  examples git:(implement_import) ✗ tf apply -auto-approve
alks_iamrole.test_role: Refreshing state... [id=TEST-DELETE]
alks_iamrole.test_role: Destroying... [id=TEST-DELETE]
alks_iamrole.test_role: Destruction complete after 1s
alks_iamrole.test_role: Creating...
alks_iamrole.test_role: Creation complete after 1s [id=TEST-DELETE]

Apply complete! Resources: 1 added, 0 changed, 1 destroyed.
```
